### PR TITLE
an alternative approach to the cleanup migration

### PR DESCRIPTION
**Pull request by hardwarehuman**
This illustrates some changes. They work, but there's other issues we need to discuss and investigate before we call it done.

### CODE REVIEW

**Review by AdamAdHocTeam — COMMENTED** [review]
It looks like this just uses a CTE instead of a temp table.

The run against the most recent db backup from yesterday was 31 minutes just for the delete.

image

File: src/migrations/20230828103020-program-personnel-cleanup.js [hunk 1]
```diff
@@ -10,35 +10,29 @@ module.exports = {
...
+         /* 1. Remove records we don't want */
+         WITH initial_entries AS (
+         SELECT
+           "firstName",
+           "lastName",
+           "role",
+           "grantId",
+           "programId",
+           active,
+           MIN(id) AS first_insert
```
Comment (GarrettEHill) [inline src/migrations/20230828103020-program-personnel-cleanup.js:22]: This is not the best set of fields to match on, I saw some records that only changed email address, while name remained the same. and others where the name changed and email remained the same. I think email needs to be included for comparison.

File: src/migrations/20230828103020-program-personnel-cleanup.js [hunk 2]
```diff
@@ -10,35 +10,29 @@ module.exports = {
...
+         FROM "ProgramPersonnel"
+         GROUP BY 1,2,3,4,5,6
          )
+         DELETE FROM "ProgramPersonnel"
+         WHERE id NOT IN (SELECT first_insert FROM initial_entries);
```
Comment (GarrettEHill) [inline src/migrations/20230828103020-program-personnel-cleanup.js:27]: Why are you using the "in select" syntax vs the using syntax

File: src/migrations/20230828103020-program-personnel-cleanup.js [hunk 3]
```diff
@@ -10,35 +10,29 @@ module.exports = {
...
+         
+         /* 2. Make a copy of the audit log before cleaning it up */
+         SET session_replication_role = replica;
+         CREATE TABLE "ZZarchiveZALProgramPersonnel20230828103020" AS SELECT * FROM "ZALProgramPersonnel";
```
Comment (GarrettEHill) [inline src/migrations/20230828103020-program-personnel-cleanup.js:31]: This is a real table and have an audit log generated for it. I don't think that is what is intended.

File: src/migrations/20230828103020-program-personnel-cleanup.js [hunk 4]
```diff
@@ -10,35 +10,29 @@ module.exports = {
...
+           active,
+           MIN(id) AS first_insert
+         FROM "ProgramPersonnel"
+         GROUP BY 1,2,3,4,5,6
          )
+         DELETE FROM "ProgramPersonnel"
+         WHERE id NOT IN (SELECT first_insert FROM initial_entries);
+         
+         /* 2. Make a copy of the audit log before cleaning it up */
+         SET session_replication_role = replica;
```
Comment (GarrettEHill) [inline src/migrations/20230828103020-program-personnel-cleanup.js:30]: This is not how to turn off audit log

File: src/migrations/20230828103020-program-personnel-cleanup.js [hunk 5]
```diff
@@ -10,35 +10,29 @@ module.exports = {
...
+         CREATE TABLE "ZZarchiveZALProgramPersonnel20230828103020" AS SELECT * FROM "ZALProgramPersonnel";
+         SET session_replication_role = DEFAULT ;
 
-         /* 3. Delete everything we don't want to keep */
-         DELETE FROM "ProgramPersonnel" WHERE id NOT IN (SELECT "idToKeep" FROM "ProgramPersonnelToKeep");
+         /* 3. Clean out the huge number of records related to deleted dupes */
+         DELETE FROM "ZALProgramPersonnel" WHERE data_id NOT IN (SELECT id FROM "ProgramPersonnel");
```
Comment (GarrettEHill) [inline src/migrations/20230828103020-program-personnel-cleanup.js:35]: Why are you using the "in select" syntax vs the using syntax

**Comment by thewatermethod** [comment]
Should we close this?

**Comment by thewatermethod** [comment]
Should we close this?

**Comment by hardwarehuman** [comment]
Yeah, this was for discussion not merging, and the outcome of the discussion has been incorporated elsewhere.

**Pull request closed without merging**